server/src/test/java/org/elasticsearch/search/internal/ContextIndexSearcherTests.java

```diff
@@ -181,220 +187,91 @@ public void collect(int doc, long bucket) throws IOException {
-
-                @Override
-                public Object reduce(Collection<Collector> collectors) throws IOException {
-                    return null;
-                }
-            });
-            LeafSlice[] leafSlices = ContextIndexSearcher.computeSlices(
-                directoryReader.getContext().leaves(),
-                executor.getMaximumPoolSize(),
-                1
-            );
-            assertEquals(leafSlices.length, collectorCalls.get());
-        } finally {
-            directoryReader.close();
-            directory.close();
-            executor.shutdown();
+            return numDocs;
         }
     }
 
-    public void testConcurrentSearchAllThreadsFinish() throws Exception {
-        final Directory directory = newDirectory();
-        IndexWriter iw = new IndexWriter(directory, new IndexWriterConfig(new StandardAnalyzer()).setMergePolicy(NoMergePolicy.INSTANCE));
-        final int numDocs = randomIntBetween(100, 200);
-        for (int i = 0; i < numDocs; i++) {
-            Document document = new Document();
-            document.add(new StringField("field", "value", Field.Store.NO));
-            iw.addDocument(document);
-            if (rarely()) {
-                iw.commit();
+    /**
+     * Check that knn queries rewrite parallelizes on the number of segments if there are enough threads in the pool.
+     */
+    public void testConcurrentRewrite() throws Exception {
+        try (Directory directory = newDirectory()) {
+            indexDocs(directory);
+            try (DirectoryReader directoryReader = DirectoryReader.open(directory)) {
+                AtomicInteger executeCalls = new AtomicInteger(0);
+                int bound = randomIntBetween(1, 50);
+                ContextIndexSearcher searcher = new ContextIndexSearcher(
+                    directoryReader,
+                    IndexSearcher.getDefaultSimilarity(),
+                    IndexSearcher.getDefaultQueryCache(),
+                    IndexSearcher.getDefaultQueryCachingPolicy(),
+                    1,
+                    randomBoolean(),
+                    new BoundedExecutor(command -> {
+                        executeCalls.incrementAndGet();
+                        command.run();
+                    }, bound),
+                    randomBoolean()
+                );
+                // check that we calculate one slice per segment
+                int numSegments = directoryReader.getContext().leaves().size();
+                assertEquals(numSegments, searcher.slices(directoryReader.getContext().leaves()).length);
+
+                KnnFloatVectorQuery vectorQuery = new KnnFloatVectorQuery("float_vector", new float[] { 0, 0, 0 }, 10, null);
+                Query rewrittenQuery = vectorQuery.rewrite(searcher);
+                // Note: we expect one execute calls less than segments since the last is executed on the caller thread.
+                // For details see QueueSizeBasedExecutor#processTask
```

Comment:
```suggestion
                // For details see BoundedExecutor#processTask
```

Reply:
this is actually correct, it refers to the executor currently used in Lucene.

Reply:
😕 my bad.